File: examples/plugin-host/host.py
# SPDX-License-Identifier: Apache-2.0
"""A plugin host, built on the Zygo API alone.

The exit criterion for the embedder API (roadmap 2.12), and it is written as a
program rather than a checklist because the question it answers is not "does
each route work" — the suites answer that — but **can somebody build the thing
the API is for without reaching past it**. No `sandbox.toml`, no file on the
Zygo host, no shell on the Zygo host. If this needed one, the API was missing a
route.

What a plugin host is: software with customers who supply code. A CI service, a
notebook backend, a webhook platform, an agent framework with tools. Each
customer's code is theirs, runs under their own limits, and must not be able to
reach anybody else's.

    the host (this file)                       zygo api
    ─────────────────────                      ────────
    onboard a customer      ──────────────►    POST /tenants, POST .../tokens
    set what they may use   ──────────────►    PATCH /tenants/<id>/limits
    give them a key         ──────────────►    PUT  /tenants/<id>/secrets/<n>
    declare one runtime     ──────────────►    POST /runtimes
    install a plugin        ──────────────►    PUT  /scripts        (as them)
    run one                 ──────────────►    POST /runtimes/<r>/call
    watch a long one        ──────────────►    ?stream=1
    stop one                ──────────────►    DELETE /requests/<key>
    bill for it             ◄──────────────    --usage-webhook

Run it against a Zygo whose API is up:

    ZYGO_API_URL=unix:///run/zygo/api.sock ZYGO_API_TOKEN=... python3 host.py
"""

from __future__ import annotations

import base64
import io
import os
import sys
import tarfile

sys.path.insert(0, os.environ.get("ZYGO_SDK", "../../sdk/python/src"))

import zygo  # noqa: E402

#: One runtime per language, and no more than that. Each holds an interpreter
#: and a dependency set and **no code at all**, which is what makes it safe for
#: several customers to share one — and why ten thousand plugins are not ten
#: thousand warm processes.
#:
#: A customer's limits are on the *customer*, not on the runtime, so a plugin
#: is held to the same memory and the same deadline whichever of these it runs
#: in. That is the property this example exists to show: a host adds a language
#: by adding a pool, and nothing about its customers changes.
RUNTIMES = {
    "python": {
        "image": os.environ.get("ZYGO_IMAGE", "python:3.12-slim"),
        "agent": "python",
    },
    "javascript": {
        "image": os.environ.get("ZYGO_NODE_IMAGE", "node:22-slim"),
        "agent": "node",
    },
}

#: What a plugin is written in, when its customer did not say. Python, because
#: that is what most plugin authors reach for.
DEFAULT_LANGUAGE = "python"


def runtime_of(language: str) -> str:
    """The pool a plugin in this language runs in."""
    if language not in RUNTIMES:
        raise ValueError(f"no runtime for {language!r}; there is {', '.join(RUNTIMES)}")
    return f"plugins-{language}"


def tar_of(files: dict[str, bytes]) -> bytes:
    """A tar in memory. A plugin's input arrives as one and leaves as one."""
    buffer = io.BytesIO()
    with tarfile.open(fileobj=buffer, mode="w") as archive:
        for name, body in files.items():
            info = tarfile.TarInfo(name)
            info.size = len(body)
            archive.addfile(info, io.BytesIO(body))
    return buffer.getvalue()


class Plugin:
    """One customer's plugin: its digest, and what it is written in.

    Two fields, because Zygo only knows about the first. A digest names bytes;
    which pool those bytes can be loaded by is the host's own record, kept in
    the same row of the same table as everything else about the plugin.
    """

    __slots__ = ("digest", "language")

    def __init__(self, digest: str, language: str) -> None:
        self.digest = digest
        self.language = language

    @property
    def runtime(self) -> str:
        return runtime_of(self.language)

    def __repr__(self) -> str:
        return f"Plugin({self.digest[:19]}…, {self.language})"


class PluginHost:
    """Everything this host knows how to do, in about a hundred lines."""

    def __init__(self, operator: zygo.Client) -> None:
        self.operator = operator
        self.tokens: dict[str, str] = {}

    # ---- the operator's side -------------------------------------------

    def start(self) -> None:
        """Declare one runtime per language every plugin may be written in.

        No `base_dir`: this host has no files on the Zygo machine, and a pool
        that named one would be the first thing to break when the API moved to
        another box.

        The limits here are the *pool's* ceiling, and are deliberately
        generous: what holds a plugin down is its customer's own limits, set in
        `onboard`, which narrow these per request. Adding a language is this
        dictionary getting one more entry.
        """
        for language, runtime in RUNTIMES.items():
            self.operator.serve_runtime(
                runtime_of(language),
                {
                    **runtime,
                    "min_warm": 1,
                    "max_warm": 8,
                    "timeout": "120s",
                    "seccomp": "strict",
                },
            )

    def onboard(self, customer: str, *, mem: str, secrets: dict[str, str]) -> str:
        """A new customer: their tenant, their limits, their keys, their token.

        The token is the only thing that leaves this method. Everything the
        customer's code can reach afterwards follows from it — which is the
        property that makes a plugin host possible at all.
        """
        self.operator.create_tenant(customer)
        self.operator.set_limits(customer, mem=mem, pids=64, timeout="60s")
        for name, value in secrets.items():
            self.operator.put_secret(customer, name, value)
        minted = self.operator.mint_token(customer)
        self.tokens[customer] = minted.secret
        return minted.secret

    def offboard(self, customer: str) -> dict:
        """Everything of theirs goes: code, tokens, secrets, running work."""
        return self.operator.delete_tenant(customer)

    def stop(self) -> None:
        """Drop every pool. The reverse of `start`."""
        for language in RUNTIMES:
            self.operator.stop_runtime(runtime_of(language))

    # ---- a customer's side ---------------------------------------------

    def client(self, customer: str) -> zygo.Client:
        """A client acting as that customer, with their own token."""
        return zygo.connect(self.operator.endpoint.url, token=self.tokens[customer])

    def install(
        self, customer: str, source: str, language: str = DEFAULT_LANGUAGE
    ) -> "Plugin":
        """Register a plugin's code against its customer, by digest.

        Sent once, called any number of times. The digest is theirs: another
        customer naming it is told the script does not exist.

        The language is the host's own bookkeeping — Zygo has no opinion about
        which pool a digest belongs in, and would happily hand a Python script
        to the Node pool, where it would fail to load. A real host keeps this
        in the same row as the plugin.
        """
        with self.client(customer) as client:
            return Plugin(client.put_script(source).sha256, language)

    def run(
        self, customer: str, plugin: "Plugin", event: dict, files: dict | None = None
    ):
        """Call a plugin, optionally with files in and files out."""
        with self.client(customer) as client:
            if files is None:
                return client.run_script(plugin.runtime, plugin.digest, event)
            return client.run_script(
                plugin.runtime,
                plugin.digest,
                event,
                workspace={"inline": base64.b64encode(tar_of(files)).decode()},
                out=True,
            )

    def watch(self, customer: str, plugin: "Plugin", event: dict, on_output) -> object:
        """Call a plugin and hand its output on as it is produced."""
        with self.client(customer) as client:
            for item in client.stream_script(plugin.runtime, plugin.digest, event):
                if item.is_result:
                    return item.result
                on_output(item.kind, item.data)
        return None

    def run_cancellable(self, customer: str, plugin: "Plugin", event: dict, key: str):
        """Call a plugin under a name this host can stop it by."""
        with self.client(customer) as client:
            return client.run_script(plugin.runtime, plugin.digest, event, key=key)

    def cancel(self, customer: str, key: str) -> dict:
        with self.client(customer) as client:
            return client.cancel(key)

File: examples/plugin-host-node/host.mjs
// SPDX-License-Identifier: Apache-2.0
/**
 * A plugin host in Node, built on the Zygo API alone.
 *
 * The Node twin of `examples/plugin-host`: software with customers who supply
 * code, where each customer's code runs under their own limits and cannot
 * reach anybody else's. No `sandbox.toml`, no file on the Zygo machine, no
 * shelling out to `zygo` — one operator token and the SDK.
 *
 * Where the Python example is a class its demo calls, this one is also an HTTP
 * server, because that is what a plugin host is to its customers:
 *
 *     a customer            this host (node:http)                zygo api
 *     ──────────            ─────────────────────                ────────
 *     PUT  /plugins    ──►  forTenant(c).putScript   ──►  PUT  /scripts
 *     POST /plugins/…/run   forTenant(c).runScript   ──►  POST /runtimes/<r>/call
 *     POST /plugins/…/stream    …streamScript        ──►  …?stream=1
 *     DELETE /runs/<key>        …cancel              ──►  DELETE /requests/<key>
 *
 *     the operator
 *     ────────────
 *     POST   /customers ──►  createTenant, setLimits, mintToken
 *     DELETE /customers/<id> ──►  deleteTenant
 *
 * Run it against a Zygo whose API is up:
 *
 *     ZYGO_API_URL=unix:///run/zygo/api.sock ZYGO_API_TOKEN=... node host.mjs
 */

import http from 'node:http';
import { pathToFileURL } from 'node:url';

// From npm this line is `from 'zygo-sdk'`; the repository's own copy is used
// here so the example runs from a checkout with nothing installed.
import { Busy, Cancelled, HandlerError, NotFound, Timeout, Unavailable, ZygoError, connect } from '../../sdk/node/src/index.js';

/**
 * One runtime pool per language, and no more than that. A pool holds an
 * interpreter and a dependency set and **no code at all**: the script arrives
 * with the request and is loaded in the forked child, after its seccomp
 * filter. That is what makes it safe for many customers to share one, and why
 * ten thousand plugins are not ten thousand warm processes.
 */
export const RUNTIMES = {
  python: { image: process.env.ZYGO_IMAGE ?? 'python:3.12-slim', agent: 'python' },
  javascript: { image: process.env.ZYGO_NODE_IMAGE ?? 'node:22-slim', agent: 'node' },
};

/** The pool a plugin in this language runs in. */
export function runtimeOf(language) {
  if (!(language in RUNTIMES)) throw new NotFound(`no runtime for ${language}; there is ${Object.keys(RUNTIMES).join(', ')}`);
  return `plugins-${language}`;
}

/** Everything this host knows how to do. */
export class PluginHost {
  /** @param {import('../../sdk/node/src/index.js').Client} operator */
  constructor(operator) {
    this.operator = operator;
    /** A customer's token, as this host recognises it: secret → customer id. */
    this.customers = new Map();
    /** Which pool a digest belongs in — the host's own record, not Zygo's. */
    this.languages = new Map();
  }

  // ---- the operator's side --------------------------------------------

  /**
   * Declare one pool per language. The limits are the pool's ceiling and are
   * generous on purpose: what holds a plugin down is its customer's own
   * limits, which narrow these on every request.
   */
  async start() {
    for (const [language, runtime] of Object.entries(RUNTIMES)) {
      await this.operator.serveRuntime(runtimeOf(language), {
        ...runtime,
        min_warm: 1,
        max_warm: 8,
        timeout: '120s',
        seccomp: 'strict',
      });
    }
  }

  /**
   * A new customer: their tenant, their limits, their token. The token is the
   * only thing that leaves; everything their code can reach follows from it.
   */
  async onboard(customer, { mem = '128M' } = {}) {
    await this.operator.createTenant(customer);
    await this.operator.setLimits(customer, { mem, pids: 64, timeout: '60s' });
    const minted = await this.operator.mintToken(customer);
    this.customers.set(minted.secret, customer);
    return minted.secret;
  }

  /** Everything of theirs goes: code, tokens, secrets, running work. */
  async offboard(customer) {
    for (const [secret, owner] of this.customers) if (owner === customer) this.customers.delete(secret);
    return this.operator.deleteTenant(customer);
  }

  /** Drop every pool. The reverse of `start`. */
  async stop() {
    for (const language of Object.keys(RUNTIMES)) await this.operator.stopRuntime(runtimeOf(language));
  }

  // ---- a customer's side ------------------------------------------------

  /**
   * The operator's client, acting for one customer. A header on the same
   * connection rather than a second client, so a thousand customers are one
   * connection pool — and scripts registered through it are theirs, and pool
   * calls may only name their own.
   */
  as(customer) {
    return this.operator.forTenant(customer);
  }

  /** Register a plugin's code against its customer, by digest. Sent once. */
  async install(customer, source, language = 'python') {
    const runtime = runtimeOf(language);
    const { sha256 } = await this.as(customer).putScript(source);
    this.languages.set(sha256, language);
    return { digest: sha256, language, runtime };
  }

  /** The pool this digest was installed into. Unknown to this host: 404. */
  poolOf(digest) {
    const language = this.languages.get(digest);
    if (!language) throw new NotFound('no such plugin');
    return runtimeOf(language);
  }

  /** Call a plugin, under a key this host can stop it by. */
  run(customer, digest, event, key) {
    return this.as(customer).runScript(this.poolOf(digest), digest, event, { key });
  }

  /** Call a plugin and yield its output as it is produced. */
  watch(customer, digest, event) {
    return this.as(customer).streamScript(this.poolOf(digest), digest, event);
  }

  cancel(customer, key) {
    return this.as(customer).cancel(key);
  }
}

// ---- the HTTP surface ---------------------------------------------------

/**
 * What a Zygo error is to a customer of this host. The host's status codes
 * are its own: a customer sees "your plugin failed", never Zygo's answer
 * passed through raw. `Busy` and `Unavailable` arrive here only after the
 * client's retries gave up, and the server's `Retry-After` is passed on.
 */
export function statusOf(error) {
  // `detail` is the exception's own text; `stderr` is what else it printed.
  if (error instanceof HandlerError) return [500, { error: 'the plugin failed', detail: error.message, stderr: error.stderr, exitCode: error.exitCode }];
  if (error instanceof Timeout) return [504, { error: 'the plugin ran out of time', detail: error.message, stderr: error.stderr }];
  if (error instanceof Cancelled) return [499, { error: 'the plugin was stopped' }];
  if (error instanceof NotFound) return [404, { error: 'no such plugin' }];
  if (error instanceof Busy || error instanceof Unavailable) {
    return [503, { error: 'try again shortly' }, { 'retry-after': String(error.retryAfter) }];
  }
  if (error instanceof SyntaxError) return [400, { error: 'the body is not JSON' }];
  if (error instanceof ZygoError) return [502, { error: 'the sandbox host refused the request' }];
  return [500, { error: 'internal error' }];
}

async function readBody(request) {
  const chunks = [];
  for await (const chunk of request) chunks.push(chunk);
  return Buffer.concat(chunks).toString('utf8');
}

function send(response, status, body, headers = {}) {
  const payload = JSON.stringify(body);
  response.writeHead(status, http.STATUS_CODES[status] ?? 'Client Closed Request', {
    'content-type': 'application/json',
    'content-length': String(Buffer.byteLength(payload)),
    ...headers,
  });
  response.end(payload);
}

/**
 * The routes, each with who may call it: the operator, or a customer whose
 * token this host minted. Every route needs a bearer token.
 */
const ROUTES = [
  ['POST', /^\/customers$/, 'operator', async (host, { body }) => {
    const { id, mem } = JSON.parse(body);
    return [201, { id, token: await host.onboard(id, { mem }) }];
  }],
  ['DELETE', /^\/customers\/([^/]+)$/, 'operator', async (host, { m }) => [200, await host.offboard(m[1])]],
  ['PUT', /^\/plugins$/, 'customer', async (host, { customer, body, url }) => [
    201,
    await host.install(customer, body, url.searchParams.get('language') ?? 'python'),
  ]],
  ['POST', /^\/plugins\/([^/]+)\/run$/, 'customer', async (host, { customer, m, body, request }) => {
    const out = await host.run(customer, m[1], JSON.parse(body || 'null'), request.headers['x-run-key']);
    return [200, { result: out.result, metrics: out.metrics }];
  }],
  ['POST', /^\/plugins\/([^/]+)\/stream$/, 'customer', async (host, { customer, m, body, response }) => {
    response.writeHead(200, { 'content-type': 'application/x-ndjson' });
    try {
      for await (const item of host.watch(customer, m[1], JSON.parse(body || 'null'))) {
        const line = item.kind === 'result' ? { result: item.result.result, metrics: item.result.metrics } : { stream: item.kind, data: item.data };
        response.write(JSON.stringify(line) + '\n');
      }
    } catch (error) {
      // The headers are gone, so the failure travels as the last line — the
      // way Zygo's own stream carries its status.
      const [status, failure] = statusOf(error);
      response.write(JSON.stringify({ status, ...failure }) + '\n');
    }
    response.end();
    return null;
  }],
  ['DELETE', /^\/runs\/([^/]+)$/, 'customer', async (host, { customer, m }) => [200, await host.cancel(customer, m[1])]],
];

/** An `http.Server` fronting the host. */
export function createServer(host) {
  return http.createServer(async (request, response) => {
    const url = new URL(request.url, 'http://host');
    const bearer = (request.headers.authorization ?? '').replace(/^Bearer /, '');
    const role = bearer !== '' && bearer === host.operator.token ? 'operator' : host.customers.has(bearer) ? 'customer' : null;
    if (!role) return send(response, 401, { error: 'no such token' });

    let m = null;
    const route = ROUTES.find(([method, pattern]) => request.method === method && (m = url.pathname.match(pattern)));
    if (!route) return send(response, 404, { error: `no route ${request.method} ${url.pathname}` });
    if (route[2] !== role) return send(response, 403, { error: `${route[2] === 'operator' ? "the operator's" : "a customer's"} token is needed here` });

    try {
      const body = await readBody(request);
      const answer = await route[3](host, { m, body, url, request, response, customer: host.customers.get(bearer) });
      if (answer) send(response, answer[0], answer[1]);
    } catch (error) {
      const [status, body, headers] = statusOf(error);
      send(response, status, body, headers);
    }
  });
}

if (process.argv[1] && import.meta.url === pathToFileURL(process.argv[1]).href) {
  const operator = connect(process.env.ZYGO_API_URL, { retries: 3 });
  const host = new PluginHost(operator);
  await host.start();
  const server = createServer(host).listen(Number(process.env.PORT ?? 8080), () => {
    console.log(`plugin host on :${server.address().port}, pools: ${Object.keys(RUNTIMES).map(runtimeOf).join(', ')}`);
  });
  process.on('SIGINT', async () => {
    server.close();
    await host.stop();
    operator.close();
  });
}
